hazelcast-client/src/test/java/com/hazelcast/client/cache/nearcache/ClientNearCacheInvalidationTest.java

```diff
@@ -0,0 +1,336 @@
+        NearCacheManager nearCacheManager = client.client.getNearCacheManager();
+        CachingProvider provider = HazelcastClientCachingProvider.createCachingProvider(client);
+        HazelcastClientCacheManager cacheManager = (HazelcastClientCacheManager) provider.getCacheManager();
+        HazelcastServerCacheManager memberCacheManager = (HazelcastServerCacheManager) HazelcastServerCachingProvider
+                .createCachingProvider(member).getCacheManager();
+
+        ICache<Object, String> cache = cacheManager.createCache(DEFAULT_CACHE_NAME, createCacheConfig(inMemoryFormat));
+        ICache<Object, String> memberCache = member.getCacheManager().getCache(
+                CacheUtil.getPrefixedCacheName(DEFAULT_CACHE_NAME, null, null));
+
+        NearCache<Data, String> nearCache = nearCacheManager.getNearCache(
+                cacheManager.getCacheNameWithPrefix(DEFAULT_CACHE_NAME));
+
+        testContext = new NearCacheTestContext(client, member, cacheManager, memberCacheManager, nearCacheManager, cache,
+                memberCache, nearCache);
+
+        // add 1 K-V pair in the cache
+        testContext.cache.put("a", "b");
+        testContext.cache.get("a");
+
+    }
+
+    @After
+    public void tearDown() {
+        hazelcastFactory.shutdownAll();
+    }
+
+    @Test
+    public void when_cacheDestroyed_invalidationEventIsReceived() {
+        final AtomicInteger counter = new AtomicInteger();
+        registerInvalidationListener(counter);
+        destroy();
+        assertInvalidationEventReceivedExactlyOnce(counter);
+    }
+
+    @Test
+    public void when_cacheCleared_invalidationEventIsReceived() {
+        final AtomicInteger counter = new AtomicInteger();
+        registerInvalidationListener(counter);
+        clear();
+        assertInvalidationEventReceivedExactlyOnce(counter);
+    }
+
+    @Test
+    public void when_cacheClosed_invalidationEventIsNotReceived() {
+        final AtomicInteger counter = new AtomicInteger();
+        registerInvalidationListener(counter);
+        close();
+        assertInvalidationEventNeverReceived(counter);
+    }
+
+    // When CacheManager.destroyCache is invoked from client-side CacheManager, an invalidation event is received
+    // When invoked from a member-side CacheManager, invalidation event is not received.
+    @Test
+    public void when_cacheManagerDestroyCacheInvoked_invalidationEventMayBeReceived() {
+        final AtomicInteger counter = new AtomicInteger();
+        registerInvalidationListener(counter);
+        destroyCacheFromCacheManager();
+        if (invokeCacheOperationsFromMember) {
+            // when from a member-side cache manager, invalidation event cannot be received, as the invalidation listener
+            // registration is removed *before* the invocation to destroy the cache is sent to the member
+            assertInvalidationEventNeverReceived(counter);
```

Comment:
See https://github.com/vbekiaris/hazelcast/blob/0deb3aed4803a85e1346f34a9a3e45c244e46bba/hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java#L235-L235

- near cache invalidation listeners are deregistered
- near cache is destroyed on the client side
- once these are done, we proceed with requesting the distributed object be destroyed https://github.com/vbekiaris/hazelcast/blob/0deb3aed4803a85e1346f34a9a3e45c244e46bba/hazelcast-client/src/main/java/com/hazelcast/client/spi/ClientProxy.java#L96-L96. The member delivers the invalidation events but they cannot be listened-for by the client, however the near cache has already been destroyed so it's not a problem.

In any case, this PR is not about _fixing_ invalidation events (one could argue that they are not needed here), it's about restoring behavior to its original state in 3.7 so we don't have any unexpected regression. We can (and should) review usages of invalidation events for 3.8.